src/crewai/rag/chromadb/utils.py

```diff
@@ -0,0 +1,221 @@
+
+
+def _is_async_client(client: ChromaDBClientType) -> TypeGuard[AsyncClientAPI]:
+    """Type guard to check if the client is an asynchronous AsyncClientAPI.
+
+    Args:
+        client: The client to check.
+
+    Returns:
+        True if the client is an AsyncClientAPI, False otherwise.
+    """
+    return isinstance(client, AsyncClientAPI)
+
+
+def _prepare_documents_for_chromadb(
+    documents: list[BaseRecord],
+) -> PreparedDocuments:
+    """Prepare documents for ChromaDB by extracting IDs, texts, and metadata.
+
+    Args:
+        documents: List of BaseRecord documents to prepare.
+
+    Returns:
+        PreparedDocuments with ids, texts, and metadatas ready for ChromaDB.
+    """
+    ids: list[str] = []
+    texts: list[str] = []
+    metadatas: list[Mapping[str, str | int | float | bool]] = []
+
+    for doc in documents:
+        if "doc_id" in doc:
+            ids.append(doc["doc_id"])
+        else:
+            content_hash = hashlib.sha256(doc["content"].encode()).hexdigest()[:16]
+            ids.append(content_hash)
+
+        texts.append(doc["content"])
+        metadata = doc.get("metadata")
+        if metadata:
+            if isinstance(metadata, list):
+                metadatas.append(metadata[0] if metadata else {})
+            else:
+                metadatas.append(metadata)
+        else:
+            metadatas.append({})
+
+    return PreparedDocuments(ids, texts, metadatas)
+
+
+def _extract_search_params(
+    kwargs: ChromaDBCollectionSearchParams,
+) -> ExtractedSearchParams:
+    """Extract search parameters from kwargs.
+
+    Args:
+        kwargs: Keyword arguments containing search parameters.
+
+    Returns:
+        ExtractedSearchParams with all extracted parameters.
+    """
+    return ExtractedSearchParams(
+        collection_name=kwargs["collection_name"],
+        query=kwargs["query"],
+        limit=kwargs.get("limit", 10),
+        metadata_filter=kwargs.get("metadata_filter"),
+        score_threshold=kwargs.get("score_threshold"),
+        where=kwargs.get("where"),
+        where_document=kwargs.get("where_document"),
+        include=kwargs.get(
+            "include",
+            [IncludeEnum.metadatas, IncludeEnum.documents, IncludeEnum.distances],
+        ),
+    )
+
+
+def _convert_distance_to_score(
+    distance: float,
+    distance_metric: Literal["l2", "cosine", "ip"],
+) -> float:
+    """Convert ChromaDB distance to similarity score.
+
+    Notes:
+        Assuming all embedding are unit-normalized for now, including custom embeddings.
+
+    Args:
+        distance: The distance value from ChromaDB.
+        distance_metric: The distance metric used ("l2", "cosine", or "ip").
+
+    Returns:
+        Similarity score in range [0, 1] where 1 is most similar.
+    """
+    if distance_metric in ("cosine", "ip"):
+        score = 1.0 - 0.5 * distance
+    else:
+        score = 1.0 - 0.25 * distance
+    return max(0.0, min(1.0, score))
```

Comment:
I liked this version, but will left my 🪙 on that

I believe the functions has issues. I'm not a mathematical expert but I got some insights from [here](https://en.wikipedia.org/wiki/Cosine_similarity) and [here](https://medium.com/@zilliz_learn/similarity-metrics-for-vector-search-62ccda6cfdd8)

Cosine: Using 0.5 * distance is incorrect. Cosine distance ranges [0,2], so it should be distance / 2.0

IP: For normalized vectors, IP similarity = 1 - distance, not 1 - 0.5 * distance

L2: Linear scaling 0.25 * distance will saturate quickly since L2 is unbounded. Consider 1.0 / (1.0 + distance) instead

So I would say you should do something like that

```suggestion
if distance_metric == "cosine":
        # Cosine distance is [0,2], divide by 2 to normalize
        score = 1.0 - (distance / 2.0)
    elif distance_metric == "ip":
        # IP distance equals (1 - similarity) for normalized vectors
        score = 1.0 - distance
    else: 
        # L2 is unbounded, use reciprocal to avoid saturation
        score = 1.0 / (1.0 + distance)
    
    return max(0.0, min(1.0, score))
```

Reply:
pushing this per internal conversations